### PR TITLE
Create github status when the build starts

### PR DESCRIPTION
**Pull request by d2lam**
This PR creates a github status when a build starts.
- Build model now takes in a 3rd parameter `password`. This is because we need to create a User model inside the constructor so that we can pass that into githubHelper. We would need to change the api to pass it in, similar to this: https://github.com/screwdriver-cd/screwdriver/blob/master/bin/server#L39
- Build create now takes in 2 extra params: config.username and config.sha: https://github.com/screwdriver-cd/screwdriver/pull/66
- Github helper has the run command, which get the user, unseal the user's token, and call github command.

### CODE REVIEW

File: lib/build.js [hunk 1]
```diff
@@ -98,26 +108,38 @@ class BuildModel extends BaseModel {
                 });
             },
             async.apply(fetchJob, this.datastore, jobId),
-            async.apply(fetchScmUrl, this.datastore)
-        ], (err, data) => {
+            async.apply(fetchScmUrl, this.datastore),
+            (data, next) => {
```
Comment (tkyi) [inline lib/build.js:110]: should this be `(err, data, next)`?

File: test/lib/github.test.js [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+const mockery = require('mockery');
+const sinon = require('sinon');
+
+sinon.assert.expose(assert, { prefix: '' });
+
+describe('Github', () => {
+    const scmUrl = 'git@github.com:screwdriver-cd/data-schema.git';
+    let schemaMock;
+    let userMock;
+    let BreakerMock;
```
Comment (tkyi) [inline test/lib/github.test.js]: make capitalization consistent

File: lib/build.js (hunk 1 above)
Comment (stjohnjohnson) [inline lib/build.js:110]: Nah, it's a waterfall.  It only gets data, next

File: lib/build.js [hunk 3]
```diff
@@ -98,26 +108,38 @@ class BuildModel extends BaseModel {
...
+                // Create github status
+                githubHelper.run({
+                    User: this.User,
+                    username: config.username,
+                    action: 'createStatus',
+                    params: {
+                        user: repoInfo.user,
+                        repo: repoInfo.repo,
+                        sha,
+                        state: 'pending'
```
Comment (stjohnjohnson) [inline lib/build.js]: Can we set the context to 'Screwdriver'.

File: lib/github.js [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    return {
+        user: matched[2],
+        repo: matched[3]
+    };
+}
+
+/**
+ * Run command. This first gets the user, then unseal the user's token and call the github command
+ * @param  {Object}   config            Configuration object
+ * @param  {Model}    config.User       User Model
```
Comment (stjohnjohnson) [inline lib/github.js]: Why is User capitalized?  It's not a class, it's an instance.

File: lib/github.js [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+            github.authenticate({
+                type: 'oauth',
+                token: unsealed
+            });
+            githubBreaker.runCommand({
+                action: config.action,
+                params: config.params || {}
+            }, next);
+        }
+    ], (err, res) => callback(err, res)
```
Comment (stjohnjohnson) [inline lib/github.js]: Why not just put callback in here?

**Comment by stjohnjohnson** [comment]
yesss
👍

File: lib/github.js [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+const async = require('async');
+const Breaker = require('circuit-fuses');
+const schema = require('screwdriver-data-schema');
+const Github = require('github');
+const github = new Github();
+let cached = null;
+
+/**
+ * Github command to run
+ * @param  {Object}   options            An object that tells what command & params to run
```
Comment (petey) [inline lib/github.js]: These are all missing the `@method` jsdoc key.

Comment (petey) [inline lib/github.js]: Might be missing a docblockr setting.
screen shot 2016-08-01 at 5 52 48 pm

Comment (d2lam) [inline lib/github.js]: Interesting. I do have that installed but it didn't enforce @method. 
I added them now.

**Comment by d2lam** [comment]
I updated the PR to expose `github.js` since we will need to use it in the `api` to lookup the `sha` for `POST /v3/builds`

@stjohnjohnson  @petey  Can you guys review again? :D

**Comment by stjohnjohnson** [comment]
Hmm, I figured you would leave the lookup to be in the models.  If the user doesn't provide a SHA, then look it up on build.create. 👍 otherwise.

**Comment by d2lam** [comment]
@stjohnjohnson  Hmm true. I could do that as well. That might actually better since the logic for the handler will be simpler. In that case I don't really need to expose github.

**Comment by d2lam** [comment]
@stjohnjohnson Let's add the lookup in the next PR since the intent of this story was just to get the Pull Request to work. I can work on that after this one is merged.

**Comment by tkyi** [comment]
👍 
giphy-1

**Pull request merged**